sdk/data/azcosmos/cosmos_location_cache.go

```diff
@@ -0,0 +1,307 @@
+	// Indicates whether multiple write locations are enabled for this database account.
+	enableMultipleWriteLocations bool
+}
+
+type locationCache struct {
+	locationInfo              dbAcctLocationsInfo
+	defaultEndpt              url.URL
+	enableEndptDiscovery      bool
+	useMultipleWriteLocations bool
+	// connLimit                         int
+	rwMutex                           sync.RWMutex
+	locationUnavailabilityInfoMap     map[url.URL]locationUnavailabilityInfo
+	mapMutex                          sync.RWMutex
+	lastUpdateTime                    time.Time
+	enableMultipleWriteLocations      bool
+	unavailableLocationExpirationTime time.Duration
+}
+
+func newLocationCache(prefLocations []string, defaultEndpt url.URL) *locationCache {
+	return &locationCache{
+		defaultEndpt:                      defaultEndpt,
+		locationInfo:                      *newDbAcctLocationsInfo(prefLocations, defaultEndpt),
+		locationUnavailabilityInfoMap:     make(map[url.URL]locationUnavailabilityInfo),
+		unavailableLocationExpirationTime: defaultExpirationTime,
+	}
+}
+
+func (lc *locationCache) update(writeLocations []AcctRegion, readLocations []AcctRegion, prefList []string, enableMultipleWriteLocations bool) error {
+	lc.rwMutex.RLock()
+	nextLoc := copyDbAcctLocationsInfo(lc.locationInfo)
+	if prefList != nil {
+		nextLoc.prefLocations = prefList
+	}
+	lc.enableMultipleWriteLocations = enableMultipleWriteLocations
+	lc.refreshStaleEndpts()
+	if readLocations != nil {
+		availReadEndptsByLocation, availReadLocations, err := getEndptsByLocation(readLocations)
+		if err != nil {
+			lc.rwMutex.RUnlock()
+			return err
+		}
+		nextLoc.availReadEndptsByLocation = availReadEndptsByLocation
+		nextLoc.availReadLocations = availReadLocations
+	}
+
+	if writeLocations != nil {
+		availWriteEndptsByLocation, availWriteLocations, err := getEndptsByLocation(writeLocations)
+		if err != nil {
+			lc.rwMutex.RUnlock()
+			return err
+		}
+		nextLoc.availWriteEndptsByLocation = availWriteEndptsByLocation
+		nextLoc.availWriteLocations = availWriteLocations
+	}
+
+	nextLoc.writeEndpts = lc.getPrefAvailableEndpts(nextLoc.availWriteEndptsByLocation, nextLoc.availWriteLocations, write, lc.defaultEndpt)
+	nextLoc.readEndpts = lc.getPrefAvailableEndpts(nextLoc.availReadEndptsByLocation, nextLoc.availReadLocations, read, nextLoc.writeEndpts[0])
+	lc.lastUpdateTime = time.Now()
+	lc.rwMutex.RUnlock()
+	lc.rwMutex.Lock()
+	lc.locationInfo = nextLoc
+	lc.rwMutex.Unlock()
+	return nil
+}
+
+func (lc *locationCache) readEndpts() ([]url.URL, error) {
+	lc.mapMutex.RLock()
+	defer lc.mapMutex.RUnlock()
+	if time.Since(lc.lastUpdateTime) > lc.unavailableLocationExpirationTime && len(lc.locationUnavailabilityInfoMap) > 0 {
+		err := lc.update(nil, nil, nil, lc.enableMultipleWriteLocations)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return lc.locationInfo.readEndpts, nil
+}
+
+func (lc *locationCache) writeEndpts() ([]url.URL, error) {
+	lc.mapMutex.RLock()
```

Comment:
Similar question as with the read endpoints, why do you need a lock?

Reply:
Just to clarify, `.Rlock()` is a read lock, which is non-blocking assuming there are no current write locks acquired. I added them to prevent any race conditions. For a `RWMutex`, you can acquire the write lock with just `.Lock()`

Reply:
My question is more around, why don't we need this on .NET?

Reply:
At the moment, there is only one map, for which I simulated the concurrent dictionary using the variable `mapMutex`. So that's why it's needed here, in another comment thread, I explained why I took this approach rather than use the sync.Map